qe/drivers.py

```diff
@@ -1707,3 +1866,682 @@ def selection_step(
+        H = self.H
+
+        return np.dot(H, V_k)
+
+
+@dataclass
+class Hamiltonian_two_electrons_integral_driven_manager(object):
+    """Instance of `Hamiltonian_two_electrons_integral_driven()` class for particular
+    lists of internal/external determinants. Used to cache `local' Hamiltonian.
+    Re-created at each CIPSI iteration."""
+
+    psi_i: Psi_det  # Internal (or local) list of determinants
+    psi_j: Psi_det  # External (or full internal) list of determinants
+    d_two_e_integral: Two_electron_integral
+
+    def __init__(
+        self,
+        psi_i: Psi_det,
+        psi_j: Psi_det,
+        d_two_e_integral: Two_electron_integral,
+        comm: MPI.COMM_WORLD = None,
+    ):
+        self.d_two_e_integral = d_two_e_integral
+        if comm is None:  # Non-MPI jobs
+            self.comm = None
+            self.psi_i = psi_i
+            self.psi_j = psi_j
+        else:  # For MPI jobs
+            self.comm = comm
+            self.size = self.comm.Get_size()  # No. of processes running
+            self.rank = self.comm.Get_rank()  # Rank of current process
+            # Dispatch wavefunction and save psi_local, dictionary of 2e integrals with local instance of class
+            if self.size == 1:  # One process running, use entire wavefunction
+                self.psi_i = psi_j
+            else:  # Else, dispatch chunks of the wavefunction to all processes
+                self.psi_i = psi_i
+            self.psi_j = psi_j
+            # Local problem dimension (size of internal wavefunction or no. of local determinants)
+            self.local_size = len(self.psi_i)
+            # Full problem dimension (no. of connected determinants or size of full internal wavefunction)
+            self.full_size = len(self.psi_j)
+            # Generate utilities
+            generator = H_indices_generator(self.psi_i, self.psi_j)
+            self.spindet_a_occ_i = generator.spindet_a_occ_int
+            self.spindet_b_occ_i = generator.spindet_b_occ_int
+            self.det_to_index_j = generator.det_to_index_ext
+
+    @cached_property
+    def Hamiltonian_two_electrons_driver(self):
+        # Create instance of driver class
+        return Hamiltonian_two_electrons_integral_driven(self.d_two_e_integral)
+
+    @cached_property
+    def H(self):
+        """Build and cache len(psi_i) \times len(psi_j) two-electron
+        Hamiltonian in an integral-driven fashion.
+        Used in building Hamiltonian to be diagonalized."""
+
+        return self.Hamiltonian_two_electrons_driver.H(self.psi_i, self.psi_j)
+
+    @cached_property
+    def H_full(self):
+        """Build full two-electron Hamiltonian H = [H_1; ...; H_N]."""
+        if self.comm is None:
+            return self.H
+        else:
+            H_i = self.H  # Each process builds local Hamiltonian
+            recvcounts = None
+            recvcounts = np.array(
+                self.comm.gather(self.local_size * self.full_size, root=0)
```

Comment:
please make a choice between `Gather` and `gather`

Reply:
I think here, the `gather` is okay, since I'm gathering integers into an array. This is what the mpi4py documentation uses in a case like this.

Reply:
Makes sense! Now it's consistent. :]